File: 02249 (Computationally Hard Problems)/Project/Code/MirrorFriendlyMinimumSpanningTree/MirrorFriendlyMinimumSpanningTree/MFMST/EdgeWeightedGraph.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using MFMSTProject.Util;

namespace MFMSTProject.MFMST {
    public class EdgeWeightedGraph : IEnumerable<Edge> {
        public int Vertices { get; private set; }
        public int EdgeCount { get { return _edges.Count; } }
        public ICollection<Edge>[] Edges { get; private set; }
        private readonly List<Edge> _edges = new List<Edge>();

        public EdgeWeightedGraph(int vertices) {
            if (vertices < 1) throw new ArgumentException("There must be at least one vertex in a graph", "vertices");
            Vertices = vertices;
            Edges = new ICollection<Edge>[vertices];
            for (var i = 0; i < vertices; i++) Edges[i] = new LinkedList<Edge>();
        }

        public void AddEdge(Edge edge) {
            Edges[edge.Vertex1].Add(edge);
            Edges[edge.Vertex2].Add(edge);
            _edges.Add(edge);
        }

        public Edge this[int index] {
            get { return _edges[index]; }
        }

        public IEnumerator<Edge> GetEnumerator() {
            for (var i = 0; i < Vertices; i++) {
                foreach (var edge in Edges[i]) {
                    if (edge.Vertex2 > i) yield return edge;
                }
            }
        }

        IEnumerator IEnumerable.GetEnumerator() {
            return GetEnumerator();
        }

        public static EdgeWeightedGraph FromStream(TextReader stream, bool manualInput) {
            var vertices = stream.ReadLine().ToInt();
            var edges = stream.ReadLine().ToInt();
            var graph = new EdgeWeightedGraph(vertices);
            var line = "";

            while ((line = stream.ReadLine()) != null) {
                var parts = line.Split(' ');
                if (parts.Length < 3) {
                    throw new ArgumentException("Each line for the edges must contain three numbers: int int int");
                };
                var vertex1 = parts[0].ToInt();
                var vertex2 = parts[1].ToInt();
                var weight = parts[2].ToInt();

                var edge = new Edge(graph.EdgeCount + 1, vertex1, vertex2, weight);
                graph.AddEdge(edge);
                if (manualInput && edges == graph.EdgeCount) break;
            }

            if (edges != graph.EdgeCount) throw new ArgumentException("Amount of edges read and the amount of edges supposed to be there does not match.");

            return graph;
        }
    }
    
}

File: 02249 - Computationally Hard Problems/Project/Code/Java/src/mfmst/EdgeWeightedGraph.java
package mfmst;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Scanner;

import util.StringFunc;

public class EdgeWeightedGraph /*: IEnumerable<Edge>*/ {
	private int _vertices;
	private LinkedList<Edge>[] _edges;
	private ArrayList<Edge> _allEdges = new ArrayList<Edge>();

    @SuppressWarnings("unchecked")
	public EdgeWeightedGraph(int vertices) {
        if (vertices < 1) throw new IllegalArgumentException("There must be at least one vertex in a graph");
        _vertices = vertices;
        _edges = (LinkedList<Edge>[]) new LinkedList[_vertices];
        for (int i = 0; i < vertices; i++) _edges[i] = new LinkedList<Edge>();
    }

    public void addEdge(Edge edge) {
        _edges[edge.getVertex1()].add(edge);
        _edges[edge.getVertex2()].add(edge);
        _allEdges.add(edge);
    }
    
    public Edge getEdge(int index) {
        return _allEdges.get(index);
    }
    
    public int getEdgeCount(){
    	return _allEdges.size();
    }
    
    public int getVertexCount(){
    	return _vertices;
    }
    
    public Iterable<Edge> getAdjacent(int vertex){
    	return _edges[vertex];
    }
    
    public static EdgeWeightedGraph fromStream(Scanner stream, boolean manualInput) {
        int vertices = stream.nextInt();
        int edges = stream.nextInt();
        EdgeWeightedGraph graph = new EdgeWeightedGraph(vertices);
        String line = "";

        while ((line = stream.nextLine()) != null) {
            String[] parts = line.split(" ");
            if (parts.length < 3) {
                throw new IllegalArgumentException("Each line for the edges must contain three numbers: int int int");
            }
            int vertex1 = StringFunc.toInt(parts[0]);
            int vertex2 = StringFunc.toInt(parts[1]);
            int weight = StringFunc.toInt(parts[2]);

            Edge edge = new Edge(graph.getEdgeCount() + 1, vertex1, vertex2, weight);
            graph.addEdge(edge);
            if (manualInput && edges == graph.getEdgeCount()) break;
        }

        if (edges != graph.getEdgeCount()) throw new IllegalArgumentException("Amount of edges read and the amount of edges supposed to be there does not match.");

        return graph;
    }
}
